### PR TITLE
CURL_LOCK_DATA_KNOWNHOSTS

### PR DESCRIPTION
**Pull request by mickae1**
When Transfering multiple file with SSH2, the access of the known_host file is not protected.
To protect the access of the file, use this option to protect the access of the file.
Be careful, only one file is authorized. Because the handle of the file is shared across connections

### CODE REVIEW

**Comment by mickae1** [comment]
@bagder  I will correct the commit, I saw your comment from the previous pull request.

**Comment by mickae1** [comment]
@bagder  
_"
I'm not sure I understand. What do you mean with "only one file is authorized" ?

Also, you should mention that this feature only works with the libssh2 backend unless you intend to fix the others as well._
"


one file is authorized. Because in the code, I have just one handle of the file known_hosts:


    if(data->share && (data->share->specifier & (1<< CURL_LOCK_DATA_HOSTS_NAME))) {
      if(data->share->known_hosts_shared == NULL)
        data->share->known_hosts_shared = ssh->kh;
      Curl_share_unlock(data, CURL_LOCK_DATA_HOSTS_NAME);
    }



I already said that its working only with libssh2:

"When Transfering multiple file with SSH2,"

**Comment by bagder** [comment]
"SSH2" is a protocol. "libssh2" is a library. As it only works with libssh2, I think it should be clear that's the case. curl can be built to use other libraries that also support SSH2.

**Comment by bagder** [comment]
@mickae1 try `make checksrc` locally yo have the code style checked:
~~~
./vssh/libssh2.c:457:83: warning: Longer than 79 columns (LONGLINE)
     if(data->share && (data->share->specifier & (1<< CURL_LOCK_DATA_KNOWNHOSTS)) )
./vssh/libssh2.c:458:81: warning: Longer than 79 columns (LONGLINE)
       Curl_share_lock(data, CURL_LOCK_DATA_KNOWNHOSTS, CURL_LOCK_ACCESS_SINGLE);
./vssh/libssh2.c:459:8: warning: Trailing whitespace (TRAILINGSPACE)
         
        ^
./vssh/libssh2.c:614:83: warning: Longer than 79 columns (LONGLINE)
     if(data->share && (data->share->specifier & (1<< CURL_LOCK_DATA_KNOWNHOSTS)) )
make[2]: *** [checksrc] Error 5
./vssh/libssh2.c:2791:100: warning: Longer than 79 columns (LONGLINE)
       if(sshc->kh && !( data->share && data->share->specifier & (1<< CURL_LOCK_DATA_KNOWNHOSTS))) {
./vssh/libssh2.c:3183:85: warning: Longer than 79 columns (LONGLINE)
     if(data->share && (data->share->specifier & (1<< CURL_LOCK_DATA_KNOWNHOSTS)) ) {
./vssh/libssh2.c:3184:81: warning: Longer than 79 columns (LONGLINE)
       Curl_share_lock(data, CURL_LOCK_DATA_KNOWNHOSTS, CURL_LOCK_ACCESS_SINGLE);
./vssh/libssh2.c:3198:84: warning: Longer than 79 columns (LONGLINE)
     if(data->share && (data->share->specifier & (1<< CURL_LOCK_DATA_KNOWNHOSTS))) {

File: lib/vssh/libssh2.c [hunk 1]
```diff
@@ -3173,6 +3179,11 @@ static CURLcode ssh_connect(struct connectdata *conn, bool *done)
 #ifdef HAVE_LIBSSH2_KNOWNHOST_API
   if(data->set.str[STRING_SSH_KNOWNHOSTS]) {
     int rc;
+
+    if(data->share && (data->share->specifier & (1<< CURL_LOCK_DATA_KNOWNHOSTS)) ) {
+      Curl_share_lock(data, CURL_LOCK_DATA_KNOWNHOSTS, CURL_LOCK_ACCESS_SINGLE);
+      ssh->kh = data->share->known_hosts_shared;
+    }
     ssh->kh = libssh2_knownhost_init(ssh->ssh_session);
```
Comment (bagder) [inline lib/vssh/libssh2.c:3187]: `ssh->kh` gets overwritten here even if assigned in the block just above. Is that really intended?

**Comment by mickae1** [comment]
> "SSH2" is a protocol. "libssh2" is a library. As it only works with libssh2, I think it should be clear that's the case. curl can be built to use other libraries that also support SSH2.

Ok I didn't know that it was possible to have other library than libSSH2.

File: lib/share.h [hunk 2]
```diff
@@ -57,6 +57,7 @@ struct Curl_share {
   struct Curl_ssl_session *sslsession;
   size_t max_ssl_sessions;
   long sessionage;
+  void *known_hosts_shared;
```
Comment (bagder) [inline lib/share.h:60]: Everything in this struct is shared, I don't think appending `_shared` to this name here makes much sense.

File: docs/libcurl/curl_share_setopt.3 [hunk 3]
```diff
@@ -94,6 +94,11 @@ The Public Suffix List stored in the share object is made available to all
 easy handle bound to the later. Since the Public Suffix List is periodically
 refreshed, this avoids updates in too many different contexts.
 
+.IP CURL_LOCK_DATA_HOSTS_NAME
+When Transfering multiple file with SSH2, the access of the known_host file is not protected. 
+To protect the access of the file, use this option to protect the access of the file
+Be careful, only one file is authorized. Because the handle of the file is shared across connections
+
```
Comment (bagder) [inline docs/libcurl/curl_share_setopt.3:101]: The rest of this file wraps the content at around column 80, maybe you should for this section as well? And let's work on the text, I really don't understand it. My attempt:

> Share the `known_hosts` handle for SFTP and SCP transfers. Only works with the libssh2 backend. Added in 7.74.0

**Review by bagder — CHANGES_REQUESTED** [review]
We lack a libcurl test case for this new functionality.

Can you also please squash all the commits into a single one, rebase it and force-push?

File: docs/libcurl/curl_share_setopt.3 (hunk 3 above)
Comment (bagder) [reply docs/libcurl/curl_share_setopt.3:100]: Also note that you added this new section *in the midddle* of the `CURL_LOCK_DATA_PSL` explanation.

Comment (mickae1) [reply docs/libcurl/curl_share_setopt.3]: ok corrected

File: lib/share.h (hunk 2 above)
Comment (mickae1) [reply lib/share.h]: ok corrected

**Comment by mickae1** [comment]
@bagder  i've squashed the commit. Everything should be ok. but I don't know yet how to create a test. I've to dig into that ....

File: docs/libcurl/curl_share_setopt.3 [hunk 4]
```diff
@@ -43,6 +43,11 @@ sure that the callback uses a different lock for each kind of data.
 \fIaccess\fP defines what access type libcurl wants, shared or single.
 
 \fIuserptr\fP is the pointer you set with \fICURLSHOPT_USERDATA\fP.
+
+.IP CURL_LOCK_DATA_KNOWNHOSTS
+Share the known_hosts handle for SFTP and SCP transfers. 
+Only works with the libssh2 backend. Added in 7.74.0
```
Comment (bagder) [inline docs/libcurl/curl_share_setopt.3:49]: `7.75.0`

File: docs/libcurl/symbols-in-versions [hunk 5]
```diff
@@ -921,6 +921,7 @@ CURL_LOCK_DATA_NONE             7.10.3
 CURL_LOCK_DATA_PSL              7.61.0
 CURL_LOCK_DATA_SHARE            7.10.4
 CURL_LOCK_DATA_SSL_SESSION      7.10.3
+CURL_LOCK_DATA_KNOWNHOSTS       7.74.0
```
Comment (bagder) [inline docs/libcurl/symbols-in-versions:924]: `7.75.0`

File: docs/libcurl/curl_share_setopt.3 [hunk 6]
```diff
@@ -43,6 +43,11 @@ sure that the callback uses a different lock for each kind of data.
 \fIaccess\fP defines what access type libcurl wants, shared or single.
 
 \fIuserptr\fP is the pointer you set with \fICURLSHOPT_USERDATA\fP.
+
+.IP CURL_LOCK_DATA_KNOWNHOSTS
+Share the known_hosts handle for SFTP and SCP transfers.
```
Comment (bagder) [inline docs/libcurl/curl_share_setopt.3:48]: Maybe elaborate slightly more on what "the known_hosts handle" is for users who don't know libcurl internals?

File: lib/vssh/libssh2.c [hunk 7]
```diff
@@ -2782,7 +2793,8 @@ static CURLcode ssh_statemach_act(struct connectdata *conn, bool *block)
 
     case SSH_SESSION_FREE:
 #ifdef HAVE_LIBSSH2_KNOWNHOST_API
-      if(sshc->kh) {
+      if(sshc->kh && !(data->share &&
+        data->share->specifier & (1<< CURL_LOCK_DATA_KNOWNHOSTS)) ) {
```
Comment (bagder) [inline lib/vssh/libssh2.c:2797]: Shouldn't the freeing of this how rather be done when the corresponding share object is freed? How do you prevent that this free isn't done while another handle us still using it?

**Review by bagder — CHANGES_REQUESTED** [review]

**Comment by bagder** [comment]
For test inspiration, check out test 506, 586, 1538, 1541 etc for libcurl-using tests that already use the share interface:

They're controlled by `tests/data/test[num]` and then the libcurl program to run is written in `tests/libcurl/lib[num].c`

File: docs/libcurl/curl_share_setopt.3 [hunk 8]
```diff
@@ -46,7 +46,10 @@ sure that the callback uses a different lock for each kind of data.
 
 .IP CURL_LOCK_DATA_KNOWNHOSTS
 Share the known_hosts handle for SFTP and SCP transfers. 
-Only works with the libssh2 backend. Added in 7.74.0
+This will share the content of the known_hosts file among multiple connection.
```
Comment (bagder) [inline docs/libcurl/curl_share_setopt.3:49]: "between multiple transfers" perhaps? Share objects are set by transfer.

File: docs/libcurl/curl_share_setopt.3 [hunk 9]
```diff
@@ -43,6 +43,14 @@ sure that the callback uses a different lock for each kind of data.
 \fIaccess\fP defines what access type libcurl wants, shared or single.
 
 \fIuserptr\fP is the pointer you set with \fICURLSHOPT_USERDATA\fP.
+
+.IP CURL_LOCK_DATA_KNOWNHOSTS
+Share the known_hosts handle for SFTP and SCP transfers. 
+This will share the content of the known_hosts file between multiple transfert.
```
Comment (danielgustafsson) [inline docs/libcurl/curl_share_setopt.3:49]: s/transfert/transfers/

File: docs/libcurl/curl_share_setopt.3 [hunk 10]
```diff
@@ -43,6 +43,14 @@ sure that the callback uses a different lock for each kind of data.
 \fIaccess\fP defines what access type libcurl wants, shared or single.
 
 \fIuserptr\fP is the pointer you set with \fICURLSHOPT_USERDATA\fP.
+
+.IP CURL_LOCK_DATA_KNOWNHOSTS
+Share the known_hosts handle for SFTP and SCP transfers. 
+This will share the content of the known_hosts file between multiple transfert. 
+It's usefull if you are usings thread to transfer data simultaneously.
```
Comment (danielgustafsson) [inline docs/libcurl/curl_share_setopt.3:50]: s/usefull/useful/
s/usings/using/
s/thread/threads/

File: docs/libcurl/curl_share_setopt.3 [hunk 11]
```diff
@@ -43,6 +43,14 @@ sure that the callback uses a different lock for each kind of data.
 \fIaccess\fP defines what access type libcurl wants, shared or single.
 
 \fIuserptr\fP is the pointer you set with \fICURLSHOPT_USERDATA\fP.
+
+.IP CURL_LOCK_DATA_KNOWNHOSTS
+Share the known_hosts handle for SFTP and SCP transfers. 
+This will share the content of the known_hosts file between multiple transfert. 
+It's usefull if you are usings thread to transfer data simultaneously. 
+Otherwise there is risk that two transfert write at the same time to the file.
```
Comment (danielgustafsson) [inline docs/libcurl/curl_share_setopt.3:51]: s/transfert/transfers/

File: docs/libcurl/symbols-in-versions (hunk 5 above)
Comment (bagder) [reply docs/libcurl/symbols-in-versions]: ~~7.75.0~~

7.77.0 perhaps?

Comment (mickae1) [reply docs/libcurl/symbols-in-versions]: I have to find time to work on it ^^ .

**Comment by bagder** [comment]
no action, change requests and now merge conflicts. Closing

**Comment by mickae1** [comment]
it's not forgotten, I just don't have time for that yet :(

**Comment by bagder** [comment]
Lovely, let me reopen this then!

**Comment by danielgustafsson** [comment]
Closing as this has stalled, please feel free to resubmit this patch either here on in a new PR when/if the above is addressed.

**Pull request closed without merging**